File: test/adaptive-viewport/validate-applied-layout-matching-test.py
#!/usr/bin/env python3

"""
Validation Script for Applied Layout Matching Property Test
Validates that the property test is correctly implemented and can run
"""

import os
import re
import sys

def main():
    print("=== Validating Applied Layout Matching Property Test ===\n")
    
    validations_passed = 0
    validations_failed = 0
    
    # Get script directory
    script_dir = os.path.dirname(os.path.abspath(__file__))
    
    # Validation 1: Check test file exists
    print("Validation 1: Checking test file exists...")
    test_file = os.path.join(script_dir, "applied-layout-matching-property.test.js")
    if os.path.exists(test_file):
        print("✓ Test file exists\n")
        validations_passed += 1
    else:
        print("✗ Test file not found\n")
        validations_failed += 1
        return validations_passed, validations_failed
    
    # Validation 2: Check HTML runner exists
    print("Validation 2: Checking HTML test runner exists...")
    html_file = os.path.join(script_dir, "test-applied-layout-matching-property.html")
    if os.path.exists(html_file):
        print("✓ HTML test runner exists\n")
        validations_passed += 1
    else:
        print("✗ HTML test runner not found\n")
        validations_failed += 1
    
    # Read test file content
    with open(test_file, 'r', encoding='utf-8') as f:
        test_content = f.read()
    
    # Validation 3: Check test file structure
    print("Validation 3: Checking test file structure...")
    
    required_elements = [
        ("Property 13 comment", r"Property 13: Applied Layout Matches Calculated Layout"),
        ("Requirements validation", r"Validates: Requirements 4\.2"),
        ("Test function export", r"runAppliedLayoutMatchingPropertyTest"),
        ("fast-check usage", r"fc\.assert"),
        ("Position tolerance constant", r"POSITION_TOLERANCE\s*=\s*1"),
        ("Property 1 test", r"Property 1:.*Single element"),
        ("Property 2 test", r"Property 2:.*Multiple elements"),
        ("Property 3 test", r"Property 3:.*Complete layout"),
        ("Property 4 test", r"Property 4:.*Multiple updates"),
        ("Property 5 test", r"Property 5:.*1px tolerance"),
        ("Position matching function", r"positionsMatch"),
        ("DOMUpdater usage", r"new DOMUpdater"),
        ("Async property tests", r"fc\.asyncProperty"),
        ("Test cleanup", r"cleanupElements")
    ]
    
    structure_valid = True
    for name, pattern in required_elements:
        if re.search(pattern, test_content):
            print(f"  ✓ {name} found")
        else:
            print(f"  ✗ {name} missing")
            structure_valid = False
    
    if structure_valid:
        print("✓ Test file structure is valid\n")
        validations_passed += 1
    else:
        print("✗ Test file structure is incomplete\n")
        validations_failed += 1
    
    # Read HTML file content
    with open(html_file, 'r', encoding='utf-8') as f:
        html_content = f.read()
    
    # Validation 4: Check HTML runner structure
    print("Validation 4: Checking HTML runner structure...")
    
    required_html_elements = [
        ("Property 13 title", r"Property 13.*Applied Layout Matches Calculated Layout"),
        ("Requirements reference", r"Requirements 4\.2"),
        ("fast-check script", r"setup-fast-check\.js"),
        ("DOMUpdater script", r"dom-updater\.js"),
        ("Test script", r"applied-layout-matching-property\.test\.js"),
        ("Run button", r'id="runTests"'),
        ("Output area", r'id="output"'),
        ("Status area", r'id="status"'),
        ("Test area", r'id="test-area"'),
        ("Test function call", r"runAppliedLayoutMatchingPropertyTest")
    ]
    
    html_valid = True
    for name, pattern in required_html_elements:
        if re.search(pattern, html_content):
            print(f"  ✓ {name} found")
        else:
            print(f"  ✗ {name} missing")
            html_valid = False
    
    if html_valid:
        print("✓ HTML runner structure is valid\n")
        validations_passed += 1
    else:
        print("✗ HTML runner structure is incomplete\n")
        validations_failed += 1
    
    # Validation 5: Check test configuration
    print("Validation 5: Checking test configuration...")
    
    config_checks = [
        ("100 iterations for Property 1", r"numRuns:\s*100.*Property 1"),
        ("100 iterations for Property 2", r"numRuns:\s*100.*Property 2"),
        ("100 iterations for Property 3", r"numRuns:\s*100.*Property 3"),
        ("50 iterations for Property 4", r"numRuns:\s*50.*Property 4"),
        ("100 iterations for Property 5", r"numRuns:\s*100.*Property 5"),
        ("60 second timeout", r"timeout:\s*60000"),
        ("1px tolerance", r"POSITION_TOLERANCE\s*=\s*1")
    ]
    
    config_valid = True
    for name, pattern in config_checks:
        if re.search(pattern, test_content, re.DOTALL):
            print(f"  ✓ {name} configured")
        else:
            print(f"  ✗ {name} not configured")
            config_valid = False
    
    if config_valid:
        print("✓ Test configuration is valid\n")
        validations_passed += 1
    else:
        print("✗ Test configuration is incomplete\n")
        validations_failed += 1
    
    # Validation 6: Check property test coverage
    print("Validation 6: Checking property test coverage...")
    
    property_tests = [
        ("Single element position matching", r"Property 1:.*Single element.*position.*match"),
        ("Batch update position matching", r"Property 2:.*Multiple elements.*batch"),
        ("Complete layout configuration", r"Property 3:.*Complete layout.*configuration"),
        ("Multiple updates accuracy", r"Property 4:.*Multiple updates.*accuracy"),
        ("Tolerance boundary test", r"Property 5:.*1px tolerance")
    ]
    
    coverage_valid = True
    for name, pattern in property_tests:
        if re.search(pattern, test_content, re.IGNORECASE):
            print(f"  ✓ {name} tested")
        else:
            print(f"  ✗ {name} not tested")
            coverage_valid = False
    
    if coverage_valid:
        print("✓ Property test coverage is complete\n")
        validations_passed += 1
    else:
        print("✗ Property test coverage is incomplete\n")
        validations_failed += 1
    
    # Validation 7: Check dependencies
    print("Validation 7: Checking dependencies...")
    
    dependencies = [
        ("DOMUpdater", "../../js/adaptive-viewport/dom-updater.js"),
        ("LayoutOptimizer", "../../js/adaptive-viewport/layout-optimizer.js"),
        ("Constants", "../../js/adaptive-viewport/constants.js"),
        ("Types", "../../js/adaptive-viewport/types.js")
    ]
    
    deps_valid = True
    for name, dep_path in dependencies:
        full_path = os.path.join(script_dir, dep_path)
        if os.path.exists(full_path):
            print(f"  ✓ {name} exists")
        else:
            print(f"  ✗ {name} not found at {dep_path}")
            deps_valid = False
    
    if deps_valid:
        print("✓ All dependencies exist\n")
        validations_passed += 1
    else:
        print("✗ Some dependencies are missing\n")
        validations_failed += 1
    
    # Print summary
    print("=" * 60)
    print("Validation Summary")
    print("=" * 60)
    print(f"Total Validations: {validations_passed + validations_failed}")
    print(f"Passed: {validations_passed}")
    print(f"Failed: {validations_failed}")
    print(f"Success Rate: {(validations_passed / (validations_passed + validations_failed) * 100):.1f}%")
    print()
    
    if validations_failed == 0:
        print("✅ All validations passed! Test is ready to run.")
        print()
        print("To run the test:")
        print("  1. Open test-applied-layout-matching-property.html in a browser")
        print("  2. Click 'Run Property Tests'")
        print("  3. Wait for all 450 iterations to complete")
        print()
        return 0
    else:
        print("❌ Some validations failed. Please fix the issues above.")
        print()
        return 1

if __name__ == "__main__":
    sys.exit(main())

File: test/adaptive-viewport/validate-applied-layout-matching-test.js
#!/usr/bin/env node

/**
 * Validation Script for Applied Layout Matching Property Test
 * Validates that the property test is correctly implemented and can run
 */

const fs = require('fs');
const path = require('path');

console.log('=== Validating Applied Layout Matching Property Test ===\n');

let validationsPassed = 0;
let validationsFailed = 0;

/**
 * Validation 1: Check test file exists
 */
console.log('Validation 1: Checking test file exists...');
const testFilePath = path.join(__dirname, 'applied-layout-matching-property.test.js');
if (fs.existsSync(testFilePath)) {
  console.log('✓ Test file exists\n');
  validationsPassed++;
} else {
  console.error('✗ Test file not found\n');
  validationsFailed++;
}

/**
 * Validation 2: Check HTML runner exists
 */
console.log('Validation 2: Checking HTML test runner exists...');
const htmlFilePath = path.join(__dirname, 'test-applied-layout-matching-property.html');
if (fs.existsSync(htmlFilePath)) {
  console.log('✓ HTML test runner exists\n');
  validationsPassed++;
} else {
  console.error('✗ HTML test runner not found\n');
  validationsFailed++;
}

/**
 * Validation 3: Check test file structure
 */
console.log('Validation 3: Checking test file structure...');
const testContent = fs.readFileSync(testFilePath, 'utf8');

const requiredElements = [
  { name: 'Property 13 comment', pattern: /Property 13: Applied Layout Matches Calculated Layout/ },
  { name: 'Requirements validation', pattern: /Validates: Requirements 4\.2/ },
  { name: 'Test function export', pattern: /runAppliedLayoutMatchingPropertyTest/ },
  { name: 'fast-check usage', pattern: /fc\.assert/ },
  { name: 'Position tolerance constant', pattern: /POSITION_TOLERANCE\s*=\s*1/ },
  { name: 'Property 1 test', pattern: /Property 1:.*Single element/ },
  { name: 'Property 2 test', pattern: /Property 2:.*Multiple elements/ },
  { name: 'Property 3 test', pattern: /Property 3:.*Complete layout/ },
  { name: 'Property 4 test', pattern: /Property 4:.*Multiple updates/ },
  { name: 'Property 5 test', pattern: /Property 5:.*1px tolerance/ },
  { name: 'Position matching function', pattern: /positionsMatch/ },
  { name: 'DOMUpdater usage', pattern: /new DOMUpdater/ },
  { name: 'Async property tests', pattern: /fc\.asyncProperty/ },
  { name: 'Test cleanup', pattern: /cleanupElements/ }
];

let structureValid = true;
requiredElements.forEach(({ name, pattern }) => {
  if (pattern.test(testContent)) {
    console.log(`  ✓ ${name} found`);
  } else {
    console.error(`  ✗ ${name} missing`);
    structureValid = false;
  }
});

if (structureValid) {
  console.log('✓ Test file structure is valid\n');
  validationsPassed++;
} else {
  console.error('✗ Test file structure is incomplete\n');
  validationsFailed++;
}

/**
 * Validation 4: Check HTML runner structure
 */
console.log('Validation 4: Checking HTML runner structure...');
const htmlContent = fs.readFileSync(htmlFilePath, 'utf8');

const requiredHTMLElements = [
  { name: 'Property 13 title', pattern: /Property 13.*Applied Layout Matches Calculated Layout/ },
  { name: 'Requirements reference', pattern: /Requirements 4\.2/ },
  { name: 'fast-check script', pattern: /setup-fast-check\.js/ },
  { name: 'DOMUpdater script', pattern: /dom-updater\.js/ },
  { name: 'Test script', pattern: /applied-layout-matching-property\.test\.js/ },
  { name: 'Run button', pattern: /id="runTests"/ },
  { name: 'Output area', pattern: /id="output"/ },
  { name: 'Status area', pattern: /id="status"/ },
  { name: 'Test area', pattern: /id="test-area"/ },
  { name: 'Test function call', pattern: /runAppliedLayoutMatchingPropertyTest/ }
];

let htmlValid = true;
requiredHTMLElements.forEach(({ name, pattern }) => {
  if (pattern.test(htmlContent)) {
    console.log(`  ✓ ${name} found`);
  } else {
    console.error(`  ✗ ${name} missing`);
    htmlValid = false;
  }
});

if (htmlValid) {
  console.log('✓ HTML runner structure is valid\n');
  validationsPassed++;
} else {
  console.error('✗ HTML runner structure is incomplete\n');
  validationsFailed++;
}

/**
 * Validation 5: Check test configuration
 */
console.log('Validation 5: Checking test configuration...');

const configChecks = [
  { name: '100 iterations for Property 1', pattern: /numRuns:\s*100.*Property 1/s },
  { name: '100 iterations for Property 2', pattern: /numRuns:\s*100.*Property 2/s },
  { name: '100 iterations for Property 3', pattern: /numRuns:\s*100.*Property 3/s },
  { name: '50 iterations for Property 4', pattern: /numRuns:\s*50.*Property 4/s },
  { name: '100 iterations for Property 5', pattern: /numRuns:\s*100.*Property 5/s },
  { name: '60 second timeout', pattern: /timeout:\s*60000/ },
  { name: '1px tolerance', pattern: /POSITION_TOLERANCE\s*=\s*1/ }
];

let configValid = true;
configChecks.forEach(({ name, pattern }) => {
  if (pattern.test(testContent)) {
    console.log(`  ✓ ${name} configured`);
  } else {
    console.error(`  ✗ ${name} not configured`);
    configValid = false;
  }
});

if (configValid) {
  console.log('✓ Test configuration is valid\n');
  validationsPassed++;
} else {
  console.error('✗ Test configuration is incomplete\n');
  validationsFailed++;
}

/**
 * Validation 6: Check property test coverage
 */
console.log('Validation 6: Checking property test coverage...');

const propertyTests = [
  { name: 'Single element position matching', pattern: /Property 1:.*Single element.*position.*match/i },
  { name: 'Batch update position matching', pattern: /Property 2:.*Multiple elements.*batch/i },
  { name: 'Complete layout configuration', pattern: /Property 3:.*Complete layout.*configuration/i },
  { name: 'Multiple updates accuracy', pattern: /Property 4:.*Multiple updates.*accuracy/i },
  { name: 'Tolerance boundary test', pattern: /Property 5:.*1px tolerance/i }
];

let coverageValid = true;
propertyTests.forEach(({ name, pattern }) => {
  if (pattern.test(testContent)) {
    console.log(`  ✓ ${name} tested`);
  } else {
    console.error(`  ✗ ${name} not tested`);
    coverageValid = false;
  }
});

if (coverageValid) {
  console.log('✓ Property test coverage is complete\n');
  validationsPassed++;
} else {
  console.error('✗ Property test coverage is incomplete\n');
  validationsFailed++;
}

/**
 * Validation 7: Check dependencies
 */
console.log('Validation 7: Checking dependencies...');

const dependencies = [
  { name: 'DOMUpdater', path: '../../js/adaptive-viewport/dom-updater.js' },
  { name: 'LayoutOptimizer', path: '../../js/adaptive-viewport/layout-optimizer.js' },
  { name: 'Constants', path: '../../js/adaptive-viewport/constants.js' },
  { name: 'Types', path: '../../js/adaptive-viewport/types.js' }
];

let depsValid = true;
dependencies.forEach(({ name, path: depPath }) => {
  const fullPath = path.join(__dirname, depPath);
  if (fs.existsSync(fullPath)) {
    console.log(`  ✓ ${name} exists`);
  } else {
    console.error(`  ✗ ${name} not found at ${depPath}`);
    depsValid = false;
  }
});

if (depsValid) {
  console.log('✓ All dependencies exist\n');
  validationsPassed++;
} else {
  console.error('✗ Some dependencies are missing\n');
  validationsFailed++;
}

/**
 * Print summary
 */
console.log('='.repeat(60));
console.log('Validation Summary');
console.log('='.repeat(60));
console.log(`Total Validations: ${validationsPassed + validationsFailed}`);
console.log(`Passed: ${validationsPassed}`);
console.log(`Failed: ${validationsFailed}`);
console.log(`Success Rate: ${((validationsPassed / (validationsPassed + validationsFailed)) * 100).toFixed(1)}%`);
console.log('');

if (validationsFailed === 0) {
  console.log('✅ All validations passed! Test is ready to run.');
  console.log('');
  console.log('To run the test:');
  console.log('  1. Open test-applied-layout-matching-property.html in a browser');
  console.log('  2. Click "Run Property Tests"');
  console.log('  3. Wait for all 450 iterations to complete');
  console.log('');
  process.exit(0);
} else {
  console.error('❌ Some validations failed. Please fix the issues above.');
  console.log('');
  process.exit(1);
}
